File: src/sleep-blocker.js
const { spawn } = require("node:child_process");

function createSleepBlocker(options = {}) {
  const platform = options.platform || process.platform;
  const spawnProcess = options.spawn || spawn;
  const command = options.command || "/usr/bin/caffeinate";
  let processHandle = null;

  function start() {
    if (platform !== "darwin" || processHandle) {
      return false;
    }

    try {
      const child = spawnProcess(command, ["-dimsu"], {
        stdio: "ignore",
        windowsHide: true
      });

      processHandle = child;

      child.once("error", () => {
        if (processHandle === child) {
          processHandle = null;
        }
      });

      child.once("exit", () => {
        if (processHandle === child) {
          processHandle = null;
        }
      });

      return true;
    } catch {
      processHandle = null;
      return false;
    }
  }

  function stop() {
    if (!processHandle) {
      return false;
    }

    const child = processHandle;
    processHandle = null;

    if (!child.killed) {
      child.kill("SIGTERM");
    }

    return true;
  }

  return {
    start,
    stop,
    isActive: () => Boolean(processHandle)
  };
}

module.exports = {
  createSleepBlocker
};

File: src-tauri/src/sleep_blocker.rs
use std::process::{Child, Command, Stdio};

#[derive(Default)]
pub struct SleepBlocker {
    child: Option<Child>,
}

impl SleepBlocker {
    pub fn start(&mut self) -> bool {
        if cfg!(not(target_os = "macos")) || self.child.is_some() {
            return false;
        }

        match Command::new("/usr/bin/caffeinate")
            .arg("-dimsu")
            .stdin(Stdio::null())
            .stdout(Stdio::null())
            .stderr(Stdio::null())
            .spawn()
        {
            Ok(child) => {
                self.child = Some(child);
                true
            }
            Err(_) => {
                self.child = None;
                false
            }
        }
    }

    pub fn stop(&mut self) -> bool {
        let Some(mut child) = self.child.take() else {
            return false;
        };

        let _ = child.kill();
        true
    }
}

impl Drop for SleepBlocker {
    fn drop(&mut self) {
        self.stop();
    }
}
